### DISCUSSION TITLE
Steps to install Marlin Binary Protocol on Debian using paukstelis OctoPrint install

### DISCUSSION THREAD

**Discussion by thisisnotfez** [mriscoc/Ender3V2S1#931] [Show and tell]
paukstelis OctoPrint on standard Debian: https://github.com/paukstelis/octoprint_install

I've been wanting to use the Firmware Updater for months, pretty much since I had discovered it. And I've only started my 3D printing journey in Thanksgiving 2022. However, OctoPrint has moved the base code to Python 3 and MBT depends on a package called heatshrink, which is not 3 friendly. I've scoured the internet and could not find any solution to this issue. So, I found my own.

I cloned the following repo into my ~/OctoPrint/Bin folder: https://github.com/trippwill/marlin-binary-protocol.git, then copied the following following items into the bin folder; binproto2, README.md, and setup.py. Once that was done, I ran the following command with the easy_install-3.9 found in the bin folder: ./easy_install-3.9 ~/OctoPrint/bin/ install.

But wait! It failed! On heatshrink! Ok, time to do some sleuthing. I opened setup.py (using your choice editor), and started messing with the requirements. Removed heatshrink2, failed. Placed wheel into the requirements, failed. Reverted back to the original, and removed the first heatshrink... success! Rebooted OctoPrint service, and the plugin can now see the package. Set the delays as found by MRiscoC here: https://github.com/OctoPrint/OctoPrint-FirmwareUpdater/issues/317. And flashed, flashing failed?

Had to power cycle the printer, but I'm happy to report that the update was actually a success!

Steps:
Log into host machine, ssh, GUI. Whatever, we just need a terminal.
cd OctoPrint/bin (default install location if installed using the above, otherwise, this is the location of your pip exec)
git clone https://github.com/trippwill/marlin-binary-protocol.git (this will make a new folder inside bin)
cd marlin-binary-protocol
cp -R marlin-binary-protocol/README.md ~/OctoPrint/bin/
cp -R setup.py ~/OctoPrint/bin/
cp -R binproto2/ ~/OctoPrint/bin/binproto2
cd ..
nano setup.py
change the following line 
    from: install_requires=["heatshrink>=0.3.2", "heatshrink2>=0.9", "pyserial>=3.4", "backports.time_perf_counter; python_version < '3.3'"],
    to: install_requires=["heatshrink2>=0.9", "pyserial>=3.4", "backports.time_perf_counter; python_version < '3.3'"],
save and exit
./easy_install-3.9 ~/OctoPrint/bin/ install
restart your octoprint instance and enjoy!

**Comment by murraythegoz** [comment]
Using the current octoprint with OctoPi custom image make installation quite different, since the easy_install is not available (and the dependencies are somehow failing), and you are running in a venv.
Furthermore, setuptools requires an update.

To enable the library on an OctoPi image (tested on OctoPi 1.1.0, OctoPrint 1.11.4 and Firmware Updater 1.14.1):

login into host machine
`cd /opt/octopi/oprint`
`. bin/activate`                                       (this activates the virtualenv)
`python -m pip install --upgrade pip`
`python -m pip install --upgrade setuptools`
`python -m pip install pyserial`
`python -m pip install heatshrink2`
`cd ~`                                                  (move back to home folder: venv will take care of proper installation)
`git clone https://github.com/trippwill/marlin-binary-protocol.git`  (this will make a new folder inside bin)
`cd marlin-binary-protocol` 
`sed -i 's/"heatshrink>=0.3.2", //g' setup.py` (this removes the dependency on heatshrink version 1, just like above)
`python setup.py install`
`deactivate` (to exit the venv environment)

notably, I found that a full reload of both raspberry pi and printer is required on my Ender3V2.
just restarting octoprint let the plugin recognize the newly installed library, but the required "BINARY_FILE_TRANSFER" attribute of the firmware is not recognized.

Besides, the maintainer of marlin-binary-protocol library is not maintaining it anymore. 
If somebody with python skills want to manage, he/she is very welcome.
